### DISCUSSION TITLE
Consider improving interaction between stop words and phrase search

### DISCUSSION THREAD

**Discussion by LukasKalbertodt** [meilisearch/product#793] [Feedback & Feature Proposal]
I just tried using stop words with Meili for the first time. My main goal was to reduce the number of match positions that would be returned by the API, as I'm using `showMatchesPosition`. But of course also to improve relevancy, search speed, indexing speed and index size.

But I noticed that stop words basically make phrase search unusable. It still works for phrases that don't have any stop words in it, like full names. So it's still useful. But as soon as the phrase contains a stop word, two things happen:

- The phrase is not highlighted in results anymore, i.e. `_matchesPosition` does not contain the match.
- Way more results are returned which do not contain the exact phrase.

Here you can see an example from our production application. Left is without stop words, right is with. The second and third result on the left have an exact match in the subtitles (just in case you were confused why they appeared). The second result on the right does not contain `moon` anywhere in the document.

<img width="49%" src="https://github.com/user-attachments/assets/76dd8f2e-fdc1-4bdb-8d41-00da00ee2b38" />
<img width="49%" src="https://github.com/user-attachments/assets/dccb14ac-c219-4405-ae0a-0d608584f8b2" />

I checked Google and there, phrase search also considers stop words. Well, at least I suspect that Google uses stop words _somehow_ when not using phrase search. Sorry for the ego-centric example, but it's the easiest I could come up with: searching for `"performs the perspective divide and uses the angular radius as field of view"` only brings up exactly one result (well, I suppose it will soon also bring up this very discussion...). And the exact phrase is highlighted in the result, including `the`. Changing the `the` to `a` does not result in any results. 

---

I have not really found a way to work around this problem. From my perspective, using stop words cripples phrase search, making it completely useless in many situations and very likely confusing users. That's not really an option for us, so it seems we cannot use stop words at all?

This seems like a suboptimal solution and I wanted to open this discussion to talk about this. Whether you have considered ways to improve the situation. I think it would be best if phrase search would still work as before, not being affected by any configured stop words. But I suspect that implementing this is not trivial?

**Comment by macraig** [comment]
Hi @LukasKalbertodt , thanks for bringing this to our attention. You make a good point about stop words affecting phrase search for your use case. We’ll bring this up with the team and share an update here. As a heads-up, this will probably happen in January since most of the team will be off for the holidays in the coming weeks.

**Reply by LukasKalbertodt** [reply]
Thanks for the quick reply and for the plans to discuss this :) Happy holidays to you all!

**Comment by ManyTheFish** [comment]
Hello @LukasKalbertodt and @macraig,

This is an old bug (it may be related to v1.0.0), and v1.12 will change the behavior without fixing the bug, it will not return documents anymore. 🤔 

The proper way of computing a phrase search should be to consider stop words as "blank space"/"joker terms" that can be replaced by any word. For instance, the phrase "tour of the moon" would be kind of replaced by "tour __ __ moon", and so, Meilisearch would try to match any document containing "tour" and "moon" separated by exactly 2 words.

Stop words are not considered searchable by Meilisearch and are completely skipped during the indexing process, that's why it would be difficult to retrieve the exact same phrase 😞

**Reply by LukasKalbertodt** [reply]
Mhhh so you are saying it's not trivial to implement, right? But do you still think it would be a good change? Do you agree with my assessment that using stop words makes phrase search mostly unusable? How much work would it be to implement this?

**Reply by ManyTheFish** [reply]
> But do you still think it would be a good change?

Yes, the current Meilisearch behavior should be considered a bug. You should be able to find documents with a phrase search even if they contain stop words.

> Do you agree with my assessment that using stop words makes phrase search mostly unusable?

Yes, you are completely right

> How much work would it be to implement this?

Basically, the [compute_phrase_docids](https://github.com/meilisearch/meilisearch/blob/022205af9024d8cde22dcdd29c1722f81f77ab0b/crates/milli/src/search/new/resolve_query_graph.rs#L187-L268) function is the one retrieving the matching documents for phrases, the issue here is that stop words are not indexed in the databases so Meilisearch can't retrieve any document from a stop word. So when the phrases are computed, [we fetch the documents containing all the words of the phrase](https://github.com/meilisearch/meilisearch/blob/022205af9024d8cde22dcdd29c1722f81f77ab0b/crates/milli/src/search/new/resolve_query_graph.rs#L196-L211) which will always return an empty list because nothing will match a stop word, so we should skip the stop words in this part.
But, the issue is not fixed for now, [the second part of the code will match every word checking if they have the expected proximity](https://github.com/meilisearch/meilisearch/blob/022205af9024d8cde22dcdd29c1722f81f77ab0b/crates/milli/src/search/new/resolve_query_graph.rs#L215-L252), in this part we have to take into account the stop words as well by considering them as "blank words".
So we have to bring the information of which word is a stop word to this function in order to process it properly.

Sorry for the delays

**Reply by LukasKalbertodt** [reply]
Thank you for that information! Good to hear that you consider this behavior a bug. Do you (as a team) have any plans to address that? And if yes, any idea when this might be fixed? Or are you rather looking for external contributors to step in here? Sorry for being the dude asking open source maintainers for free work, I just want to assess the situation and make informed decisions or maybe even contributing dev time :)

**Reply by ManyTheFish** [reply]
For now,
I don't think we have the time to address this issue, so if we want to have this issue fixed as soon as possible, we may rely on external work,
however, this is not a bug we can ignore for too long so we'll have to address it in the next months.

Sorry for my elusive answer, it's hard for me to do any premise 😞

Let's create a bug issue on the repository for better visibility.

> related to https://github.com/meilisearch/meilisearch/issues/5298

**Reply by ManyTheFish** [reply]
Hello @LukasKalbertodt,
I've tried a fix for this issue, I put below a prototype image containing the fix:

```
$ docker pull getmeili/meilisearch:prototype-fix-phrase-stop-words-00
```

Could you try it on your end and give me some feedback?

Thank you!
